rs/cli/src/qualification/mod.rs

```diff
@@ -0,0 +1,458 @@
+    step_range: String,
+    deployment_name: String,
+    prometheus_endpoint: String,
+}
+
+impl QualificationExecutorBuilder {
+    pub fn new(dre_ctx: DreContext) -> Self {
+        Self {
+            dre_ctx,
+            from_version: "".to_string(),
+            to_version: "".to_string(),
+            step_range: "".to_string(),
+            deployment_name: "".to_string(),
+            prometheus_endpoint: "".to_string(),
+        }
+    }
+
+    pub fn with_from_version(self, from_version: String) -> Self {
+        Self { from_version, ..self }
+    }
+
+    pub fn with_to_version(self, to_version: String) -> Self {
+        Self { to_version, ..self }
+    }
+
+    pub fn with_step_range(self, step_range: String) -> Self {
+        Self { step_range, ..self }
+    }
+
+    pub fn with_deployment_namge(self, deployment_name: String) -> Self {
+        Self { deployment_name, ..self }
+    }
+
+    pub fn with_prometheus_endpoint(self, prometheus_endpoint: String) -> Self {
+        Self { prometheus_endpoint, ..self }
+    }
+
+    pub fn build(self) -> QualificationExecutor {
+        QualificationExecutor::_new(self)
+    }
+}
+
+impl QualificationExecutor {
+    fn _new(ctx: QualificationExecutorBuilder) -> Self {
+        let steps = vec![
+            // Blessing the version which we are qualifying
+            Steps::EnsureBlessedVersions(EnsureBlessedRevisions {
+                version: ctx.to_version.clone(),
+            }),
+            // Upgrading deployment canisters
+            Steps::UpgradeDeploymentCanisters(UpgradeDeploymentCanisters {}),
+            // Upgrading all application subnets
+            Steps::UpgradeSubnets(UpgradeSubnets {
+                action: Action::Upgrade,
+                subnet_type: Some(SubnetType::Application),
+                to_version: ctx.to_version.clone(),
+            }),
+            // Upgrading all system subnets
+            Steps::UpgradeSubnets(UpgradeSubnets {
+                action: Action::Upgrade,
+                subnet_type: Some(SubnetType::System),
+                to_version: ctx.to_version.clone(),
+            }),
+            // Upgrading unassigned nodes
+            Steps::UpgradeSubnets(UpgradeSubnets {
+                action: Action::Upgrade,
+                subnet_type: None,
+                to_version: ctx.to_version.clone(),
+            }),
+            // Run workload tests
+            Steps::RunWorkloadTest(Workload {
+                version: ctx.to_version.clone(),
+                deployment_name: ctx.deployment_name.clone(),
+                prometheus_endpoint: ctx.prometheus_endpoint.clone(),
+            }),
+            // Run XNet tests
+            Steps::RunXnetTest(RunXnetTest {
+                version: ctx.to_version.clone(),
+            }),
+            // Since the initial testnet is spunup with disk-img
+            // retire the initial version.
+            Steps::RetireBlessedVersions(RetireBlessedVersions {
```

Comment:
i wouldn't bother with retiring versions

Reply:
Its mandatory for the qualification to work because of how testnets are bootstrapped.